online/src/main/scala/ai/chronon/online/Fetcher.scala

```diff
@@ -190,23 +190,40 @@ class Fetcher(val kvStore: KVStore,
             val baseMap = internalMap ++ externalMap
             // used for derivation based on ts/ds
             val tsDsMap: Map[String, AnyRef] =
-              Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
-            val derivedMap: Map[String, AnyRef] = Try(
+              Map("ts" -> requestTs, "ds" -> requestDs)
+                .mapValues(_.asInstanceOf[AnyRef])
+            val derivedMapTry: Try[Map[String, AnyRef]] = Try {
               joinCodec
                 .deriveFunc(internalResponse.request.keys, baseMap ++ tsDsMap)
                 .mapValues(_.asInstanceOf[AnyRef])
-                .toMap) match {
-              case Success(derivedMap) => derivedMap
-              case Failure(exception) => {
-                ctx.incrementException(exception)
-                throw exception
+            }
+
+            val (derivedMapCleaned, baseMapCleaned) = if (derivedMapTry.isSuccess) {
+              val derivedMap = derivedMapTry.get -- tsDsMap.keys
+              (derivedMap, baseMap)
+            } else {
+              ctx.incrementException(derivedMapTry.failed.get)
+              val derivedExceptionMap = Map("derivation_fetch_exception" -> derivedMapTry.failed.get.traceString)
+                .mapValues(_.asInstanceOf[AnyRef])
+              val joinTry: Try[JoinOps] = getJoinConf(joinName)
+              if (joinTry.isFailure) {
+                // joinTry can only be successful if the code can reach here
+                ctx.incrementException(joinTry.failed.get)
+                (Map.empty[String, AnyRef], Map.empty[String, AnyRef])
+              } else {
+                val join = joinTry.get
+                if (join.derivationsContainStar && !join.areDerivationsRenameOnly) {
+                  // if it is not rename only with a star in it, we need to return the base map
+                  (Map.empty[String, AnyRef], baseMap)
+                } else {
+                  (derivedExceptionMap, Map.empty[String, AnyRef])
+                }
```

Comment:
@nikhilsimha  This is the code block when the derivation try is failed. Can you take a look?

Reply:
cc: @hzding621 can you also take a look here? Thanks!